src/docs/arch-fast-truncate.dox

```diff
@@ -0,0 +1,847 @@
+structure, add a \c WT_TXN_OP_REF_DELETE operation to the current transaction (except if
+we're truncating the history store, which is non-transactional), and set the ref state to
+\c WT_REF_DELETED.
+
+Otherwise we report back to the tree-walk code that we couldn't delete the page and it
+needs to visit it.
+(This happens by a return flag and not by returning an error.)
+
+Note that it is not possible to fast-truncate an already deleted page; ordinarily the
+tree-walk code will have already skipped over it (see @ref ft_page_skip) and also, \c
+__wt_delete_page won't accept one.
+If one truncate reaches a page truncated by another transaction that is not yet visible,
+such that the skip code doesn't skip it, we need to load the page, instantiate it, and
+attempt to slow-truncate it; this will discover the transaction conflict.
+
+Also note that the first and last pages of a truncate operation are always slow-truncated
+regardless of eligibility; this is a result of the initial positioning of the start and
+end cursors, which requires the pages under them to be present in memory.
+This point is not particularly important operationally but can create complications for
+writing tests.
+
+Finally note that currently the initial eviction attempt is done unconditionally even in
+cases where we could determine beforehand that the page will be ineligible.
+This causes it to be evicted and read in again, which is suboptimal, and should perhaps be
+improved at some point.
+
+@section ft_deleted_generation Generation of other deleted pages
+
+As mentioned earlier, there are four ways besides fast-truncate that deleted pages can
+appear.
+
+First, the checkpoint cleanup code in \c bt_sync.c discards pages that it finds contain
+only obsolete values, instead of writing them out.
+The ref state becomes \c WT_REF_DELETED, no \c page_del is generated, and when the
+checkpoint reaches the parent internal page any prior on-disk page image will be dropped
+and no cell will be produced.
+
+Second, pages that reconcile empty end up deleted.
+During a checkpoint this happens in the checkpoint cleanup; in eviction, it happens in the
+\c WT_PM_REC_EMPTY case of \c __evict_page_dirty_update.
+The ref state becomes \c WT_REF_DELETED and no \c page_del is generated.
+(Also note that \c WT_REF_MEM pages with \c WT_PM_REC_EMPTY reconciliation results are
+explicitly skipped during internal page reconciliation.)
+
+In VLCS, empty deleted pages are inserted during certain internal page reads; see @ref
+ft_vlcs.
+
+Finally, new trees are created with a single empty deleted leaf page, because creating the
+tree with no leaves at all causes problems.
+
+@section ft_page_skip Skipping deleted pages
+
+There are two separate skip functions for skipping deleted pages during tree walks.
+The basic skip function is \c __wt_delete_page_skip in \c bt_delete.c; it is always called
+in tree walks when \c WT_READ_SKIP_DELETED is set, which is all tree walks except RTS.
```

Comment:
Can we point to the RTS page `arch-rts`?

Reply:
Good thought. Done in [af26591](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/af265919038ffe7086fbdc1db36c6800c63bc466)